k2/torch/csrc/fsa_class.cu

```diff
@@ -0,0 +1,476 @@
+ * limitations under the License.
+ */
+
+#include <exception>
+#include <string>
+#include <vector>
+
+#include "k2/csrc/device_guard.h"
+#include "k2/csrc/fsa_algo.h"
+#include "k2/csrc/fsa_utils.h"
+#include "k2/csrc/ragged_ops.h"
+#include "k2/torch/csrc/deserialization.h"
+#include "k2/torch/csrc/fsa_class.h"
+#include "k2/torch/csrc/utils.h"
+
+namespace k2 {
+
+FsaClass::FsaClass(const std::string &s,
+                   const std::vector<std::string> &extra_label_names /*= {}*/) {
+  // TODO: pass following options from arguments
+  bool openfst = false;
+  int32_t num_extra_labels = 0;
+  Array2<int32_t> extra_labels;
+  Array2<int32_t> *p_extra_labels;
+  int32_t num_ragged_labels = 0;
+  Ragged<int32_t> *ragged_labels = nullptr;
+
+  if (!extra_label_names.empty()) {
+    num_extra_labels = extra_label_names.size();
+    p_extra_labels = &extra_labels;
+  }
+
+  fsa = FsaFromString(s, openfst, num_extra_labels, p_extra_labels,
+                      num_ragged_labels, ragged_labels);
+
+  if (num_extra_labels) {
+    for (int32_t i = 0; i != num_extra_labels; ++i) {
+      const auto &name = extra_label_names[i];
+      Array1<int32_t> row = extra_labels.Row(i);
+      tensor_attrs[name] = Array1ToTorch(row);
+      all_attr_names.insert(name);
+    }
+  }
+  // Check the validation of this fsa, will trigger a fatal error if this fsa
+  // is not valid.
+  Properties();
+
+  // TODO: we also need to pass the name of ragged_labels.
+}
+
+FsaClass::FsaClass(const Ragged<Arc> &fsa, torch::Tensor aux_labels)
+    : fsa(fsa) {
+  K2_CHECK_EQ(fsa.NumElements(), aux_labels.numel());
+  K2_CHECK_EQ(aux_labels.scalar_type(), torch::kInt32);
+  K2_CHECK(ContextFromTensor(aux_labels)->IsCompatible(*fsa.Context()));
+  SetTensorAttr("aux_labels", aux_labels);
+}
+
+FsaClass::FsaClass(const Ragged<Arc> &fsa, Ragged<int32_t> &aux_labels)
+    : fsa(fsa) {
+  K2_CHECK_EQ(fsa.NumElements(), aux_labels.Dim0());
+  K2_CHECK(IsCompatible(fsa, aux_labels));
+  SetRaggedTensorAttr("aux_labels", aux_labels);
+}
+
+FsaClass FsaClass::FromUnaryFunctionTensor(FsaClass &src,
+                                           const Ragged<Arc> &arcs,
+                                           torch::Tensor arc_map) {
+  FsaClass dest(arcs);
+  // Check the validation of the fsa, will trigger a fatal error if the fsa
+  // is not valid.
+  dest.Properties();
+  dest.CopyTensorAttrs(src, arc_map);
+  dest.CopyRaggedTensorAttrs(src, arc_map);
+  return dest;
+}
+
+FsaClass FsaClass::FromUnaryFunctionRagged(FsaClass &src,
+                                           const Ragged<Arc> &arcs,
+                                           Ragged<int32_t> &arc_map) {
+  FsaClass dest(arcs);
+  // Check the validation of the fsa, will trigger a fatal error if the fsa
+  // is not valid.
+  dest.Properties();
+  for (const auto &iter : src.tensor_attrs) {
+    if (iter.second.scalar_type() == torch::kInt32) {
+      torch::Tensor value = iter.second.clone();
+      auto masking =
```

Comment:
For decoding, we don't need to check for `-1` in the arc map, as that will never happen.